### PR TITLE
core: correctly truncate unicode strings

### LINKED ISSUE #14897: PSI: Oops! Something went wrong. unicode truncation json parse problem
### FAQ

- [X] Yes, my issue is not about [variability](https://github.com/GoogleChrome/lighthouse/blob/main/docs/variability.md) or [throttling](https://github.com/GoogleChrome/lighthouse/blob/main/docs/throttling.md).
- [X] Yes, my issue is not about a specific accessibility audit (file with [axe-core](https://github.com/dequelabs/axe-core) instead).

### URL

https://www.kissmyparcel.com/retail/

### What happened?

Simply tried checking page speed score, got "Oops something went wrong" message.

### What did you expect?

Get the score.

### What have you tried?

_No response_

### How were you running Lighthouse?

web.dev

### Lighthouse Version

10.0.2

### PR DESCRIPTION
**Pull request by connorjclark**
In various places, we truncate web strings using `.slice`, which can result in eliding in the middle of a grapheme (visual character). This is typically fine for JSON serialization (it just won't look great), but in PSI there is a serialization step that breaks under this input.

This PR improves all the places (but maybe I missed some?) that truncate strings that end up in the LHR to instead elide using `Intl.Segmenter`. It also changes the result a bit, by considering the suffix added to the truncated strings as part of the max character length - and also, using an ellipse in some places that instead used `...`.

reprise of #11697 #11698

fixes #14897

### CODE REVIEW

File: core/lib/page-functions.js [hunk 1]
```diff
@@ -132,10 +142,9 @@ function getOuterHTMLSnippet(element, ignoreAttrs = [], snippetCharacterLimit =
       }
 
       // Elide attribute value if too long.
-      if (attributeValue.length > ATTRIBUTE_CHAR_LIMIT) {
-        attributeValue = attributeValue.slice(0, ATTRIBUTE_CHAR_LIMIT - 1) + '…';
-        dirty = true;
-      }
+      const truncatedString = truncate(attributeValue, ATTRIBUTE_CHAR_LIMIT);
+      dirty ||= truncatedString !== attributeValue;
```
Comment (adamraine) [inline core/lib/page-functions.js:146]: nit: `||=` is hard to understand here IMO

```suggestion
      if(!dirty) dirty = (truncatedString !== attributeValue);
```

File: core/audits/byte-efficiency/unminified-javascript.js [hunk 2]
```diff
@@ -8,6 +8,7 @@ import {ByteEfficiencyAudit} from './byte-efficiency-audit.js';
 import * as i18n from '../../lib/i18n/i18n.js';
 import {computeJSTokenLength as computeTokenLength} from '../../lib/minification-estimator.js';
 import {getRequestForScript, isInline} from '../../lib/script-helpers.js';
+import { Util } from '../../../shared/util.js';
```
Comment (adamraine) [inline core/audits/byte-efficiency/unminified-javascript.js:11]: nit
```suggestion
import {Util} from '../../../shared/util.js';
```

File: core/computed/unused-css.js [hunk 3]
```diff
@@ -101,16 +101,16 @@ class UnusedCSS {
           firstRuleStart > firstRuleEnd ||
           firstRuleStart > PREVIEW_LENGTH) {
         // We couldn't determine the first rule-set or it's not within the preview
-        preview = preview.slice(0, PREVIEW_LENGTH) + '...';
+        preview = Util.truncate(preview, PREVIEW_LENGTH);
       } else if (firstRuleEnd < PREVIEW_LENGTH) {
         // The entire first rule-set fits within the preview
-        preview = preview.slice(0, firstRuleEnd + 1) + ' ...';
+        preview = preview.slice(0, firstRuleEnd + 1) + ' …';
```
Comment (adamraine) [inline core/computed/unused-css.js:107]: Why not use `Util.truncate(0, firstRuleEnd, ' …')` here?

File: core/lib/lh-error.js [hunk 4]
```diff
@@ -107,14 +107,14 @@ const str_ = i18n.createIcuMessageFn(import.meta.url, UIStrings);
...
- * @typedef {{sentinel: '__LighthouseErrorSentinel', code: string, stack?: string, [p: string]: string|undefined}} SerializedLighthouseError
+ * @typedef {{sentinel: '__LighthouseErrorSentinel', code: string, stack?: string, [p: string]: unknown}} SerializedLighthouseError
  * @typedef {{sentinel: '__ErrorSentinel', message: string, code?: string, stack?: string}} SerializedBaseError
  */
 
 class LighthouseError extends Error {
   /**
    * @param {LighthouseErrorDefinition} errorDefinition
-   * @param {Record<string, string|undefined>=} properties
+   * @param {Record<string, unknown>=} properties
```
Comment (adamraine) [inline core/lib/lh-error.js:117]: Why is this change necessary?

File: core/lib/page-functions.js [hunk 5]
```diff
@@ -78,6 +80,14 @@ function getElementsInDocument(selector) {
   return results;
 }
 
+/**
+ * @param {string} string
+ * @param {number} characterLimit
+ */
+function truncate(string, characterLimit) {
+  return Util.truncate(string, characterLimit);
```
Comment (adamraine) [inline core/lib/page-functions.js:88]: `Util` isn't injected into our execution context though. Shouldn't we be using `Util.truncate.toString()` somewhere?

File: core/lib/page-functions.js (hunk 1 above)
Comment (connorjclark) [reply core/lib/page-functions.js:140]: I have the opposite opinion here 🤷

File: core/computed/unused-css.js (hunk 3 above)
Comment (connorjclark) [reply core/computed/unused-css.js:107]: Because `firstRuleEnd` is a specific location in the string, we should not truncate by characters here.

**Comment by brendankenny** [comment]
My only concern is that `Intl.Segmenter` is quite slow compared to iterating on string code points. It'll also have the _very_ slow ICU startup cost (like what necessitated [caching `Intl.NumberFormat`ers](https://github.com/GoogleChrome/lighthouse/blob/db62a55150f73afbf430eb387d338dd607fa1d6c/report/renderer/i18n-formatter.js#L21)), but hopefully that's being caught by the V8 caching of ICU init with basic options?

So for that cost, do we really need this for the handful of cases where grapheme clusters are significant ([thanks Mathias for the term](https://mathiasbynens.be/notes/javascript-unicode#other-grapheme-clusters))? Most unicode will be fine, this is mostly down in html attributes and other non-prominent strings, and it'll just be strings like `👨‍👨‍👦‍👦` that _might_ end split in the middle if they happen to end up at the length limit (though still on unicode code points, so it won't have any issue with bad string encoding).

File: core/lib/lh-error.js (hunk 4 above)
Comment (connorjclark) [reply core/lib/lh-error.js:117]: initially I needed this change when the base tsconfig target was updated to 2022, reverted in https://github.com/GoogleChrome/lighthouse/pull/14911/commits/de6a3944ae73a6b27b12e23059728461f214b642 . the base tsconfig didn't need the change anymore after I moved the truncate function to `shared/util.js` (it _was_ in `page-functions.js`). Might as well keep this more correct type, though.

**Comment by paulirish** [comment]
> My only concern is that `Intl.Segmenter` is quite slow compared to iterating on string code points. It'll also have the _very_ slow ICU startup cost (like what necessitated [caching `Intl.NumberFormat`ers](https://github.com/GoogleChrome/lighthouse/blob/db62a55150f73afbf430eb387d338dd607fa1d6c/report/renderer/i18n-formatter.js#L21)), but hopefully that's being caught by the V8 caching of ICU init with basic options?

i was skeptical too. numberformat has been a pain for us.

I was able to capture a profile of just axe + getnodedetails (in this branch):   https://trace.cafe/t/Rd8274lf8y
fwiw: getnodedetails ran on 39 elems in this case (https://www.kissmyparcel.com/retail/)
I saw truncate take 14ms on the first time, but then it was fast on all subsequent invocations. Also, on my second/third traces i couldn't repro this large instantiate cost, but I did on my fourth. shrug. some weird caching presumably.

i made this snippet to just test from within devtools: https://gist.github.com/paulirish/e46ec350be36cd23047c350602c47435

given what i'm seeing here.. so far I feel okay about the perf side.  14ms isn't nothing, but.. it seems like it's a one-time cost?

**Comment by brendankenny** [comment]
> given what i'm seeing here.. so far I feel okay about the perf side. 14ms isn't nothing, but.. it seems like it's a one-time cost?

That could be the ICU startup time, and if that is the case, seems like V8 is caching it for us, which is great. We should be wary of ever doing this for the first time during tracing.

If it looks good on perf, great to be doing the best segmenting here!

**Review by adamraine — APPROVED** [review]

File: core/lib/url-utils.js [hunk 6]
```diff
@@ -138,7 +138,7 @@ class UrlUtils {
   static elideDataURI(url) {
     try {
       const parsed = new URL(url);
-      return parsed.protocol === 'data:' ? url.slice(0, 100) : url;
+      return parsed.protocol === 'data:' ? Util.truncate(url, 100) : url;
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply core/lib/url-utils.js:141]: Good point, but at least inclusion of the ellipse suffix is an improvement here.

**Pull request merged**